blueprint/core/blueprint_ViewManager.cpp

```diff
@@ -52,14 +52,22 @@ namespace blueprint
     {
         const auto& [view, shadow] = getViewHandle(viewId);
 
-        view->setProperty(name, value);
-        shadow->setProperty(name, value);
-
-        // For now, we just assume that any new property update means we
-        // need to redraw or lay out our tree again. This is an easy future
-        // optimization.
-        performRootShadowTreeLayout();
-        view->repaint();
+        //TODO: There is an element of duplicate work checking ShadowView::isLayoutProperty and
```

Comment:
Good point... I wonder if there's a better way?

It's been a minute since I've been in this part of the codebase, but I wonder if we could have a `std::map<juce::Identifier, YogaFlexEnumValue>` so that we could have a constant lookup time to map the property name to a Yoga enum. If the map doesn't contain the identifier, we know it's not a layout prop. That, I think, reduces 2N comparisons down to the cost of the map lookup.

As for maybe returning a bool, I think we can actually do a couple more things here. So a `ShadowView` is in a lot of ways just a C++ wrapper around a YogaNode. That YogaNode has its own flag for indicating whether it's dirty or not (i.e. has a property changed since the last time we computed a layout). So (1) we could add a method on ShadowView to query the dirty state of that internal yoga node. Then (2) the way I've been doing this so far is if I set _any_ property, I redo a layout from the root of the application. Obviously we're adding the check here for `isLayoutProperty`, but it's still the root tree node. Yoga's fast, and caches well, so that's not really a problem, but in truth we often only need to do a layout from some node deeper in the tree, and only considering a much smaller subtree.

I'm thinking that after React is done applying all property updates for a given batch, we run through the whole list of shadowViews with a "if you're dirty, redo your layout for your subtree." This way the only penalty we pay is if we layout a particular subtree and then layout a parent of that subtree, but here we still get yoga's caching.

If we can do that, we can avoid this whole idea about AsyncUpdater and coalescing commits... the trick will be getting React to tell us it's done with the given batch.

What do you think? Want to try for it?

Reply:
Hi,

Thinking about what @nick-thompson suggested, using `std::map<juce::Identifier, YogaFlexEnumValue>`, and looking at the property validations / setters, does it not make sense to create a map to individual setter functions?  Then you get faster than current linear-time property setter lookups, with validation (which looks essentially like property name to setter routing) for free via std::map bounds checking.

Something like https://github.com/JoshMarler/blueprint/compare/improve_shadowview_performance...olidacombe:improve_shadowview_performance__pre_master_merge_compare 

Note the working branch is https://github.com/olidacombe/blueprint/tree/improve_shadowview_performance but that has master merged into it so the diff was pretty mental, so I made the branch above to be a readable set of changes.

I encapsulated the map into the class `PropertySetterMap` because I have designs on solving https://github.com/nick-thompson/blueprint/issues/13 as trivial side effect (in the `PropertySetterMap` constructur, copy all keys in the map with hyphens to camelCase keys with the same value).

And I've worked with the idea of returning a bool from `setProperty`, but ultimately think the search for dirty roots on end of updates batch is a great idea.

Reply:
@olidacombe, 

Been away from this for a while as other priorities took over but yeah nice! Map to setters looks great. I think you'd still want `isLayoutProp` or equivalent to `ShadowView::setProperty` (at whatever node/subtree level this occurs in), so that exceptions out of the map really are exceptional and not fired for every user defined prop etc. 

Also, I didn't think `juce::Identifier` has a `std::hash` specialisation for that map to work? @nick-thompson and I have actually had a few conversations about whether `juce::Identifier` is the right thing to use for `View` props. Every time a prop update comes in from JS/React we'll be constructing `juce::Identifier` instances on the fly (not an intended use pattern). So if you do have to switch that map to `juce::String` keys it might not be the end of the world. 

I've got this on my TODO list after a couple other things but if you want to carry on from where you've left off don't let me stop you! Looks like you're heading down a nice track.

Reply:
Cool, thanks for having a look.  When I get some time I'll crack on then with all of the above in mind.

Reply:
Interesting, nice one! I love the map of functions idea, that seems very clean. Looking over it again, I think we could likely reduce all of the preprocessor macro code there to template functions, yea? If we can do that, I think we'd clean the code up even further (wish I had thought of this ages ago hah)

Re: #13– I think we should do that on the javascript side. In general, my philosophy is to write as little C++ as possible, but write it in a way such that it is maximally flexible and maximally stable. The reason for that is that iterating on the javascript side is much faster and much easier, and would require no build changes from consumers (like SOUL, where we want to be able to ship new features and updates to SOUL end users without requiring SOUL to pull + rebuild a new binary, then requiring all SOUL users to update their install)

Reply:
Yeah I don't love the macros, especially not being able to condition on 0 variadic args, so will make a pass with variadic templates at some point.

As for https://github.com/nick-thompson/blueprint/issues/13 I wrote [babel-plugin-transform-property-cases](https://github.com/olidacombe/babel-plugin-transform-property-cases) to do this on javascript side at compile time (e.g. usage https://github.com/nick-thompson/blueprint/compare/master...olidacombe:kebabinate, annoyingly needs `transformRotate` manually mapped because `transform-rotate` is not css), but don't think it's worthy for inclusion in examples because there may be hidden pitfalls (say if someone iterates though some `{top: 5, botton: 6}` object to dynamically build `marginTop, marginBottom`, the transform can't help you).  It's fine if you know the limitations though.  Anyway, that's why I had designs on doing it C++ side at initialization because that would be fast.  Perfectly happy to abandon that idea.

Side note for @JoshMarler looking back at the map usage you get a single O(logN) lookup without throwing via `.find` and checking the iterator, so will move over to that instead of a validation pass followed by the lookup.  I was just being a rookie there.

Reply:
Awesome, thanks!

Right, yea, I think we will probably want to update `transform-rotate` anyways to be more at parity with CSS. That is, instead of supporting a `transform-rotate` property, support a `transport` property with syntax like `transform: rotate(90deg);`. I think trying to do this at compile time is likely harder than doing it at runtime (and same thing for the camel case transform). For example, we can still write our native property names with hyphens (which seems to be helpful with Yoga's enum to string helper), but catch and transform property names at the javascript layer before it reaches native right here: https://github.com/nick-thompson/blueprint/blob/master/packages/juce-blueprint/src/lib/BlueprintBackend.ts#L94. Similarly, we can implement what I'll call "macro" properties right there.

For example:
```js
  setProperty(propKey: string, value: any): any {
    this._props = Object.assign({}, this._props, {
      [propKey]: value,
    });

    if (propKey === 'transform') {
      return handleTransformMacro(value);
    }

    .... etc

    // Note the prop key transformation here
    return __BlueprintNative__.setViewProperty(this._id, camelToHyphen(propKey), value);
  },

  handleTransformMacro(value) {
    let transformList = value.split(' ');
    transformList.forEach((t) => {
      // Regex to match over e.g. rotate(90deg) such that
      // transformFunction = "rotate";
      // transformValue = 90;
      this.setProperty(`transform-${transformFunction}`, transformValue);
    });
  }
```

So you can see here that as long as our native properties support the necessary transforms, we can always compose their behavior into macro functions that support a more familiar API for the end user.

Generally I think this is the place and the strategy we'll want to use to providing a good user experience in terms of view props.

Reply:
Cool, I'm happy to do that, and will open a PR soon.  Runtime in javascript was bottom of my list for performance reasons but I see your reasoning.  It's also just easier 👍

Reply:
I get that, but I favor the simplicity :) And I suspect the runtime cost here will be insignificant, but if it's not, `shouldComponentUpdate` is more likely the correct optimization

Reply:
I've made a pass at ditching those macros.  I still used a couple of short ones to append `Percent` and `Auto` to symbols, but it's a bit cleaner now.  Also cut out the step of constructing a `YGValue`, since it didn't seem necessary.

The diff looks like this https://github.com/olidacombe/blueprint/compare/improve_shadowview_performance...olidacombe:improve_shadowview_performance-macro_strip

The `PropertySetterMap::K` typedef is ready to flip over to `juce::String` as @JoshMarler suggests if we want to do that.

Reply:
I also had a look at how to trim calls to `performRootShadowTreeLayout()`.  Do you think `resetAfterCommit` is the right signal to trigger a traverse for dirty yoga nodes (e.g. https://github.com/olidacombe/blueprint/compare/improve_shadowview_performance-macro_strip...olidacombe:resetAfterCommit-hook)?